### ISSUE TITLE
context creation takes ~2s with pipewire (~1ms with pulse)

### ISSUE THREAD

**Issue by fabiangreffrath** [kcat/openal-soft#930]
Hi,

this is in a GNOME environment on a system running recent Debian testing [1].

`$ cat oal.c`
```C
#include <stddef.h>
#include <stdio.h>
#include <time.h>

#include "alc.h"

struct timespec a, b;

int main (void)
{
    ALCdevice *device;
    ALCcontext *context;

    device = alcOpenDevice(NULL);

    clock_gettime(CLOCK_MONOTONIC_RAW, &a);
    context = alcCreateContext(device, NULL);
    clock_gettime(CLOCK_MONOTONIC_RAW, &b);

    printf("%lu\n", (b.tv_sec - a.tv_sec) * 1000000 + (b.tv_nsec - a.tv_nsec) / 1000);

    return 0;
}
```

`CFLAGS="-I/usr/include/AL -lopenal" LDLIBS=$CFLAGS make oal`

This is what happens with the pipewire backend:

```shell
$ cat alsoft.conf 
[general]
drivers = pipewire
```

```shell
$ for i in 1 2 3; do ./oal ; done
2013660
2013418
2007904
```

Now let's change to the pulse backend:

```shell
$ cat alsoft.conf 
[general]
drivers = pulse
```

```shell
$ for i in 1 2 3; do ./oal ; done
8287
9012
6866
```

[1] library versions:

```shell
ii  gstreamer1.0-pipewire:amd64     0.3.83-1        amd64        GStreamer 1.0 plugin for the PipeWire multimedia server
ii  libcanberra-pulse:amd64         0.30-10         amd64        PulseAudio backend for libcanberra
ii  libpipewire-0.3-0:amd64         0.3.83-1        amd64        libraries for the PipeWire multimedia server
ii  libpipewire-0.3-common          0.3.83-1        all          libraries for the PipeWire multimedia server - common files
ii  libpipewire-0.3-dev:amd64       0.3.83-1        amd64        libraries for the PipeWire multimedia server - development
ii  libpipewire-0.3-modules:amd64   0.3.83-1        amd64        libraries for the PipeWire multimedia server - modules
ii  libpulse-dev:amd64              16.1+dfsg1-2+b1 amd64        PulseAudio client development headers and libraries
ii  libpulse-mainloop-glib0:amd64   16.1+dfsg1-2+b1 amd64        PulseAudio client libraries (glib support)
ii  libpulse0:amd64                 16.1+dfsg1-2+b1 amd64        PulseAudio client libraries
ii  pipewire:amd64                  0.3.83-1        amd64        audio and video processing engine multimedia server
ii  pipewire-alsa:amd64             0.3.83-1        amd64        PipeWire ALSA plugin
ii  pipewire-audio                  0.3.83-1        all          recommended set of PipeWire packages for a standard audio desktop use
ii  pipewire-bin                    0.3.83-1        amd64        PipeWire multimedia server - programs
ii  pipewire-pulse                  0.3.83-1        amd64        PipeWire PulseAudio daemon
```

**Comment by kcat** [comment]
Are you able to provide an OpenAL Soft trace log? Set the `ALSOFT_LOGLEVEL` env var to `3`, and it will print to stderr.

The 2 second delay makes it seem like `PipeWirePlayback::start` is timing out while waiting to get the stream buffering info (it starts playback, then waits for about 2 seconds while periodically calling `pw_stream_get_time_n` to get the buffering length and update size for the stream, which isn't otherwise available right away). It can take a few cycles before that info becomes available, but definitely shouldn't take 2 seconds.

I'm currently using PipeWire 0.3.80 without a problem, so I wonder if some behavior changed in a newer version or there's something about the system/audio device causing the detection to fail.

**Comment by fabiangreffrath** [comment]
Sure, please find its output attached. The 2s delay happens after the `Fixed device latency: 0ns` has been printed.

```
$ ALSOFT_LOGLEVEL=3 ./oal 
[ALSOFT] (II) Initializing library v1.23.1-unknown UNKNOWN
[ALSOFT] (II) Supported backends: pipewire, pulse, alsa, sndio, oss, jack, port, null, wave
[ALSOFT] (II) Loading config /etc/openal/alsoft.conf...
[ALSOFT] (II) Loading config /etc/xdg/alsoft.conf...
[ALSOFT] (II) Loading config /home/fabian/.alsoftrc...
[ALSOFT] (II) Loading config /home/fabian/.config/alsoft.conf...
[ALSOFT] (II) Got binary: "/home/fabian", "oal"
[ALSOFT] (II) Loading config /home/fabian/alsoft.conf...
[ALSOFT] (II)  found 'drivers' = 'pipewire'
[ALSOFT] (II) Key disable-cpu-exts not found
[ALSOFT] (II) Vendor ID: ""
[ALSOFT] (II) Name: "12th Gen Intel(R) Core(TM) i5-1235U"
[ALSOFT] (II) Extensions: +SSE +SSE2 +SSE3 +SSE4.1
[ALSOFT] (II) Key rt-prio not found
[ALSOFT] (II) Key rt-time-limit not found
[ALSOFT] (II) Key game_compat/enable-sub-data-ext not found
[ALSOFT] (II) Key game_compat/reverse-x not found
[ALSOFT] (II) Key game_compat/reverse-y not found
[ALSOFT] (II) Key game_compat/reverse-z not found
[ALSOFT] (II) Key game_compat/nfc-scale not found
[ALSOFT] (II) Key resampler not found
[ALSOFT] (II) Key uhj/decode-filter not found
[ALSOFT] (II) Key uhj/filter not found
[ALSOFT] (II) Key uhj/encode-filter not found
[ALSOFT] (II) Key trap-al-error not found
[ALSOFT] (II) Key trap-alc-error not found
[ALSOFT] (II) Key reverb/boost not found
[ALSOFT] (II) Found drivers = "pipewire"
[ALSOFT] (II) Found PipeWire version "0.3.83" (0.3.80 or newer)
[ALSOFT] (II) Key pipewire/assume-audio not found
[ALSOFT] (II) Ignoring metadata "settings"
[ALSOFT] (II) Ignoring node class Midi/Bridge
[ALSOFT] (II) Ignoring metadata "route-settings"
[ALSOFT] (II) Ignoring node class Video/Source
[ALSOFT] (II) Initialized backend "pipewire"
[ALSOFT] (II) Added "pipewire" for playback
[ALSOFT] (II) Added "pipewire" for capture
[ALSOFT] (II) Key excludefx not found
[ALSOFT] (II) Key default-reverb not found
[ALSOFT] (II) Opening default playback device
[ALSOFT] (II) Got default playback device "alsa_output.pci-0000_00_1f.3.analog-stereo"
[ALSOFT] (II) Got default capture device "alsa_input.pci-0000_00_1f.3.analog-stereo"
[ALSOFT] (II) Got sink device "alsa_output.pci-0000_00_1f.3.analog-stereo"
[ALSOFT] (II)   "Built-in Audio Analog Stereo" = ID 54
[ALSOFT] (II) Device ID 54 sample rate: 48000 (range: 44100 -> 48000)
[ALSOFT] (II) Device ID 54 got 2 positions for Stereo
[ALSOFT] (II) Got source device "alsa_input.pci-0000_00_1f.3.analog-stereo"
[ALSOFT] (II)   "Built-in Audio Analog Stereo" = ID 55
[ALSOFT] (II) Device ID 55 sample rate: 48000 (range: 44100 -> 192000)
[ALSOFT] (II) Device ID 55 got 2 positions for Stereo
[ALSOFT] (II) Created device 0x55de8d4f3db0, "Built-in Audio Analog Stereo"
[ALSOFT] (II) Key frequency not found
[ALSOFT] (II) Key period_size not found
[ALSOFT] (II) Key periods not found
[ALSOFT] (II) Key sample-type not found
[ALSOFT] (II) Key channels not found
[ALSOFT] (II) Key ambi-format not found
[ALSOFT] (II) Key hrtf not found
[ALSOFT] (II) Key stereo-encoding not found
[ALSOFT] (II) Pre-reset: Stereo, Float32, 48000hz, 960 / 2880 buffer
[ALSOFT] (II) Key pipewire/rt-mix not found
[ALSOFT] (II) Post-reset: Stereo, Float32, 48000hz, 960 / 2880 buffer
[ALSOFT] (II) Key stereo-mode not found
[ALSOFT] (II) Key cf_level not found
[ALSOFT] (II) Stereo rendering
[ALSOFT] (II) Channel config, Main: 3, Real: 2
[ALSOFT] (II) Allocating 5 channels, 20480 bytes
[ALSOFT] (II) Enabling single-band first-order ambisonic decoder
[ALSOFT] (II) Key sources not found
[ALSOFT] (II) Key sends not found
[ALSOFT] (II) Max sources: 256 (255 + 1), effect slots: 64, sends: 2
[ALSOFT] (II) Key dither not found
[ALSOFT] (II) Key dither-depth not found
[ALSOFT] (II) Dithering disabled
[ALSOFT] (II) Key output-limiter not found
[ALSOFT] (II) Output limiter disabled
[ALSOFT] (II) Fixed device latency: 0ns
[ALSOFT] (II) Post-start: Stereo, Float32, 48000hz, 960 / 2880 buffer
[ALSOFT] (II) Increasing allocated voices to 256
[ALSOFT] (II) Key volume-adjust not found
[ALSOFT] (II) Created context 0x55de8d579c00
```

**Comment by kcat** [comment]
That definitely indicates it's timing out waiting for the buffer info, especially given it didn't change the "Post-start" format from "Post-reset" (PipeWire will generally make the update size a power of 2, with the buffer size some multiple of it).

The actual loop that waits for the information is [here](https://github.com/kcat/openal-soft/blob/1.23.1/alc/backends/pipewire.cpp#L1627). The only things that could be different/missing would be:

1. PipeWire is no longer setting a `spa_io_rate_match` on the stream (`mRateMatch`), or is setting one but its `rate` field is left as `0` indicating no update size. And/or,
2. The `pw_time`'s `avail_buffers` and `queued_buffers` fields are `0`, indicating no buffers available or queued. And/or,
3. The `pw_time`'s `rate` field has a `0` denominator, indicating no sample rate/speed.

If all of those are set properly, it should set the appropriate OpenAL device fields and break out of waiting. It's not good for any of that stuff to be missing, so something is behaving badly with PipeWire.

**Comment by kcat** [comment]
It seems to be a PipeWire issue, that should hopefully be resolved in the next release:
https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/3592
https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/3d8c7c40b5cc16eaf7bd1fb72c17783ce42e2d0e

It seems that sometimes, there will be an extra buffer that's not reported as queued or available, and with real-time mixing there may only be one buffer. So in some cases, `avail_buffers` and `queued_buffers` are left as 0 so OpenAL Soft can't calculate the buffer size.

The next release of PipeWire should hopefully fix it. Though in the mean time, you can disable real-time mixing which should make PipeWire use an additional buffer that OpenAL Soft can then count. Create or modify `~/.config/alsoft.conf` and add:
```
[pipewire]
rt-mix = false
```
or run `alsoft-config` and under Backends->PipeWire, deselect "Real-time Mixing" and click Apply. This will slightly increase latency, but shouldn't be any worse than other backends.

**Comment by kcat** [comment]
PipeWire 0.3.84 was released. Are you able to test if it still happens with the newest version?

**Comment by fabiangreffrath** [comment]
I'll check it out once it hits Debian testing (it's curvilinear l currently in unstable).

**Comment by fabiangreffrath** [comment]
Yes, I can confirm that this issue is fixed with PipeWire 0.3.84. Thank you very much!
